### PR TITLE
Export content items based on Query results (Lombiq Technologies: OCORE-98)

### LINKED ISSUE #11405: Ability to export content items in bulk, but not all from a given type
### Is your feature request related to a problem? Please describe.
Currently, you can either export a single content item (either directly with Export to Deployment Target or from a Content Item Step), or all content items of the given content type (with the Content Step). There is no way though to export multiple, selected content items at once.

### Describe the solution you'd like
A way to implemented this would be to utilize Queries with a "Queried Content Items" Deployment Step: It would utilize an existing Query to select matching content items, and would include those in the export. Then, in the Query, you could have any arbitrary selection criteria, even dynamic ones. (Perhaps we'd need the ability to inject parameters into the query from the Deployment Step?)

### Describe alternatives you've considered
A bulk action in the content item list. This would be a useful feature if you only want to export a handful of selected items just once.

### PR DESCRIPTION
**Pull request by lampersky**
fixes #11405

### CODE REVIEW

**Comment by lampersky** [comment]
query

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:10]:
```suggestion
<span class="hint">@T["The name of the Query which will be used to fetch content items."]</span>
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:18]:
```suggestion
    <div class="alert alert-warning">@T["If checked, you will have to manually declare variables for  content item references like Content Picker Fields, Taxonomy Fields, Tags Fields, and List item IDs and replace them where needed in the recipe. Any reference to these content items will be lost."]</div>
```

Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:18]: Why is this though? We can otherwise export such relationships.

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:23]: This doesn't look like a dropdown for me in latest Chrome, no arrow:
image

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:31]:
```suggestion
    <span class="hint">@T["The name of the Query which will be used to fetch content items. NOTE: The Query must be configured to return Documents."]</span>
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:37]:
```suggestion
    <span class="hint">@T["An optional JSON object containing the parameter values for this query."]</span>
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStepDriver.cs:32]: Add a validation error if the Query doesn't return documents.

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/_ViewImports.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/_ViewImports.cshtml:17]: Why are these changes needed?

**Review by Piedone — CHANGES_REQUESTED** [review]
Please keep comments, except for code suggestions that you apply, open, don't resolve them, so I'll be able to go through them next time.

File: src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStep.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStep.cs:9]: @Piedone please note that I'm producing already existing `content` step, which is already described here:

image

There is no need to create a distinct step.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStep.cs]: Please document this feature somewhere else, then.

File: src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentSource.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentSource.cs]: I mistakenly assumed this was supposed to be for SQL queries only.

File: src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentStepDriver.cs]: done

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: fixed

File: src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentSource.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Sql/Deployment/QueryBasedContentDeploymentSource.cs]: Due to the abstraction, it should be straightforward to make it compatible with any Query. Since a site may use Lucene, SQL, or any future/custom type of Queries, that would be quite useful.

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/_ViewImports.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/_ViewImports.cshtml]: I did it because I wanted to avoid repeating those lines inside a few **cshtml** files.
But once you've asked I realized that those are only 3 files, so I moved those usings.

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: fixed

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: I've added this because I wanted to be consistent with already existing steps.

You can read more here: https://github.com/OrchardCMS/OrchardCore/pull/6372

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: Can we make the note as info alert that show up above as we normally did

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: not sure what you are requesting as we have an alert:

image

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: Simply add an info alert similar to what you did in Export As Setup recipe, or simply surround the note with strong tag. My intend to make the note catch the eyes, coz the hint is grayed

Hope it's clear

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: I think you are mixing two different things **field hint** and **field validation**.
Everywhere in Orchard Core, field hints are grayed:
image

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: If you looked into the pic in https://github.com/OrchardCMS/OrchardCore/pull/12045#discussion_r928124991, the note is grayed like other text, probably no one can recognize there's a very importnant note about return result, so my intend to make it bold, then anyone can recognize it

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml:32]: image

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: No need to captalize NOTE,  you can use Note with surrounded strong or b tag. One more thing can we move it above where the notification showed up, or let's hear @Piedone feedback

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: While I suggested it, now that we have validation, yeah, perhaps having "NOTE" as all caps is unnecessary.

File: src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Views/Items/QueryBasedContentDeploymentStep.Fields.Edit.cshtml]: Please don't resolve such convos.

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs:21]:
```suggestion
        public QueryBasedContentDeploymentStepDriver(
            IQueryManager queryManager,
            IStringLocalizer<QueryBasedContentDeploymentStepDriver> stringLocalizer)
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs:50]: Use descriptive variable names.

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs:58]:
```suggestion
                    updater.ModelState.AddModelError(Prefix, nameof(step.QueryName), S["Your Lucene query is not returning content items."]);
                }
                else if (query.Source == "Sql" && !query.ReturnDocuments)
                {
                    updater.ModelState.AddModelError(Prefix, nameof(step.QueryName), S["Your SQL query is not returning documents."]);
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs:50]:
```suggestion
            if (await updater.TryUpdateModelAsync(model, Prefix, x => x.QueryName, x => x.QueryParameters, x => x.ExportAsSetupRecipe))
            {
```

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs:30]: Let's see if we can find a better solution than using `dynamic` and then tying the implementation to the Lucene and SQL Query types here and in the driver.

I don't have a better suggestion that to try to run the query and see what it returns, but it might not return anything by design. Perhaps we need to refactor `Query` and add a property to signal what `ReturnContentItems` and `ReturnDocuments` does (though these two are not the same indeed).

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs:30]: Perhaps @sebastienros you have a better idea, since `Query` was originally written by you?

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: Still ;).

**Comment by Piedone** [comment]
Apart from [this](https://github.com/OrchardCMS/OrchardCore/pull/12045#discussion_r928324172) and [this](https://github.com/OrchardCMS/OrchardCore/pull/12045#discussion_r928323426) LGTM, please re-request review when ready.

Please keep comments, except for code suggestions that you apply, open, don't resolve them, so I'll be able to go through them next time.

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: I've checked in many other places, and it looks like everywhere I've found same variable names. That's why I've decided to go with same short variable names here, just to be consistent. But, I will change it to be more descriptive.

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: looks like this code above is outdated,

image

@Piedone  which variable is still not descriptive enough?

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs:30]: > I don't have a better suggestion that to try to run the query and see what it returns, but it might not return anything by design. Perhaps we need to refactor `Query` and add a property to signal what `ReturnContentItems` and `ReturnDocuments` does (though these two are not the same indeed).

because we have this warning message, saying that user needs to make sure that the Query is returning content items, we could also remove those additional checks and avoid using dynamic in those two places

another idea, is to perform some additional checks inside **QueryBasedContentDeploymentSource.cs** when there is at least one item returned by the Query, and check if this item looks like **ContentItem** type and only add it if condition is met

image

I'm waiting for Sebastien's thoughts.

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: `x`

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: what would you suggest? I used same variable name as it was used in other steps.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: Well, what are those variables storing? :) Are the three `x` variables the same thing?

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs:30]: The driver, unfortunately, has the same issue, so we need to figure out a way to get rid of this tight coupling altogether.

Yeah, that checking the return object idea is what I mentioned above and I don't think is good.

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: I expect Zoltan refers to `x` which is make sense ;)

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: > Are the three `x` variables the same thing?

exactly @Piedone , this **x** is the same thing, it is an array of strongly typed lambda expressions, here it is **IUpdateModel** interface where you can see there is a **params** keyword:

image

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: we did this in many places in Orchard:

image

That's why I wasn't sure what do you mean by **Use descriptive variable names** 😛

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: In many places in Orchard, `x` doesn't tell you what's happening either ;).

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: so, what would you suggest @Piedone?

x here is related to expression, which should be included, I will change it to whatever you will prefer :)

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentStepDriver.cs]: `viewModel` on line 48, `model` here?

**Comment by lampersky** [comment]
all code suggestions applied, the only thing which is left here is https://github.com/OrchardCMS/OrchardCore/pull/12045#discussion_r928324172

**Comment by Piedone** [comment]
OK, please let me know when that's done too.

**Comment by lampersky** [comment]
> OK, please let me know when that's done too.

maybe instead of waiting for Sebastien input I will refactor Query and will add additional property, which could be used here?

**Comment by lampersky** [comment]
I think **ReturnContentItems** could be moved from **LuceneQuery** to **Query**.
Imo, it is rather a common thing to return **ContentItems** by various Query implementations.

image

**Comment by Piedone** [comment]
@Skrypt since this is affected by and affects https://github.com/OrchardCMS/OrchardCore/pull/11052, can you please advise on [this](https://github.com/OrchardCMS/OrchardCore/pull/12045#discussion_r928324172) and [this](https://github.com/OrchardCMS/OrchardCore/pull/12045#issuecomment-1213232685)?

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs:30]: I understand that Query can be inherited by SqlQuery and LuceneQuery and ElasticsearchQuery. But here you want to export content items based on SQL Queries so why not using : 

```
var query = await _queryManager.GetQueryAsync(contentStep.QueryName)

if (typeof(SqlQuery) == query.GetType())
{
   //do something
}
```

And if the casted query doesn't contain any

**Comment by Skrypt** [comment]
To be honest I'm not sure if we want to be able to ReturnContentItems from every type of Query this is why I left it for each inherited Query type to decide.

**Comment by Skrypt** [comment]
If I'm comparing each of these `Query` they also have in common `Template` but have a different name for 

ReturnContentItems == ReturnDocuments

So, if we change the type then you possibly break the Query definitions in the database. This means that you need to cast per Query to get the proper field of these classes.

**Comment by Skrypt** [comment]
After thinking. What I suggest is that we make a different PR that would assess these changes and migration. But discuss with @sebastienros about this first.

**Comment by jtkech** [comment]
@Skrypt @Piedone @lampersky About the `Query` abstraction concern.

Maybe metadata, marker interfaces, virtual methods.

For example with a new virtual method to not break existing json settings.

In the base class `Query`

        public virtual bool ResultsOfType<T>() => typeof(T) == typeof(object);

Then for example in `LuceneQuery`

        public override bool ResultsOfType<T>() => ReturnContentItems
            ? typeof(T) == typeof(ContentItem)
            : base.ResultsOfType<T>();

The usage being

            var query = await _queryManager.GetQueryAsync(id);
            if (query.ResultsOfType<ContentItem>())
            {

            }

**Comment by Piedone** [comment]
Ah, that sounds nice, yes!

File: src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs]: I've just changed PR title, because I see why you are confused. When I've opened this PR, I thought this will be only used with SQL Queries, but then Piedone told me it should work with any type of query which still returns content items.

**Comment by lampersky** [comment]
> Ah, that sounds nice, yes!

@Piedone suggestion applied

**Comment by Piedone** [comment]
Let me know if you're ready for another review.

**Comment by lampersky** [comment]
> Let me know if you're ready for another review.

I'm ready. I can't request review, because PR is already awaiting review from you.
@Piedone

**Review by Piedone — APPROVED** [review]

**Comment by sebastienros** [comment]
I don't think the changes on the interface are required. When the step is executed we can just check that it is a content item. And the description of the steps can mention that the query needs to return content items or it won't be used. Optionally it could also handle strings (ids) and do the lookup.

**Pull request merged**
